geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/MetadataFileSystemStorage.scala

```diff
@@ -153,18 +164,28 @@ abstract class MetadataFileSystemStorage(conf: Configuration,
       logger.error(s"Failed to delete some files: [${failures.mkString(", ")}]")
     }
 
-    logger.debug(s"Updating metadata for type ${metadata.getSchema.getTypeName}")
-    metadata.replaceFiles(partition, toCompact.map(_.getName).asJava, dataPath.getName)
+    logger.debug("Compacting metadata")
+
+    metadata.compact()
 
     logger.debug(s"Compacted $written records into file $dataPath")
   }
 
-  override def updateMetadata(): Unit = {
-    profile {
-      val fc = metadata.getFileContext
-      val scheme = metadata.getPartitionScheme
-      metadata.setFiles(StorageUtils.partitionsAndFiles(fc, metadata.getRoot, scheme, extension, cache = false))
-    } ((_, time) => logger.debug(s"Metadata update took ${time}ms."))
+  class AddCallback(partition: String, file: Path) extends WriterCallback {
+    override def onClose(count: Long, bounds: Envelope): Unit =
+      metadata.addPartition(new PartitionMetadata(partition, Collections.singletonList(file.getName), count, bounds))
```

Comment:
As a sanity check, we are assuming that the metadata can always be compacted into one file, right?

Reply:
no, there will be 1-n files per partition. Metadata compaction will reduce that down to one per partition

Reply:
as a follow on, it might be good to put metadata files into separate folders so we don't have too many files in one dir. I opened a ticket for that: https://geomesa.atlassian.net/browse/GEOMESA-2415